Lsig Web Support Tutorial (#851)
scale-it/algo-builder: docs/tutorials/t-10.md

# Algo Builder Web Support for Logic Signature Transaction

The purpose of this tutorial is to demonstrate the usage of Algo Builder Web Package to sign and send a Logic Signature Transaction using major Algorand wallets (Algo Signer, Wallet Connect, MyAlgoWallet).

### Requirements

- [Algo Builder](https://algobuilder.dev/) and it's [dependencies](https://algobuilder.dev/guide/README#requirements)
- Setup the accounts on [`Algosigner`](https://github.com/PureStake/algosigner), [`Wallet Connect`](https://walletconnect.com/) wallet and [`MyAlgo Wallet`](https://wallet.myalgo.com/). It could be MainNet, TestNet or private net.

## Steps

> In this tutorial we will use Algo Builder Wallet Widget to connect to various web wallets. Then we will show how to instrument the wallet to sign a multisignature transaction. See [Wallet Widget tutorial](https://developer.algorand.org/solutions/vuejs-multi-wallet-widget/) for more details about the widget.

Reviewer: wrong copy paste. Please update this section accordingly.

Author: oh yes, will update

### LogicSignature Transaction

To create a lsig transaction, you can follow steps from the from the offical documentation: [Create Logic Signature Transaction](https://developer.algorand.org/docs/get-details/dapps/smart-contracts/frontend/smartsigs/).
Here is a glimpse of creating the lsig transaction:

```js
let fs = require("fs");
let data = fs.readFileSync(filePath);
let results = await algodclient.compile(data).do();
let program = new Uint8Array(Buffer.from(results.result, "base64"));
let lsig = new algosdk.LogicSigAccount(program);

let lsigTxn = algosdk.makePaymentTxnWithSuggestedParams(
	lsig.address(),
	receiver,
	amount,
	closeToRemaninder,
	note,
	params
);
```

After creating the transaction, you can sign it using different wallets.

> 1. AlgoSigner<br/>
>    We use `connectAlgoSigner()` method to connect the AlgoSigner wallet. Next, we create an instance of the `WebMode` class by passing `AlgoSigner` and the Chain Name i.e the network name. The `web` instance will allow us to sign and send transactions to the blockchain.

Reviewer: ```suggestion
#### AlgoSigner

Use `connectAlgoSigner()` method to connect the AlgoSigner wallet. Next, we create an instance of the `WebMode` class by passing `AlgoSigner` and the Chain Name i.e the network name. The `web` instance will allow us to sign and send transactions to the blockchain.
```

```js
const web = new WebMode(AlgoSigner, CHAIN_NAME);
```

> Webmode method `signLogicSignatureTxn` which is used to sign the txn, expects transaction and LogicSigAccount as parameters.
> To sign the lsig txn:

Reviewer: ```suggestion
Use `web.signLogicSignatureTxn` method to sign the txn. It expects transaction and LogicSigAccount as arguments:
```

```js
const signedTxnResponse = await web.signLogicSignatureTxn(lsigTxn, lsig);
```

> After signing, the returned response is the object containing blob key, which we can further use to send the signed transaction to the network.

Reviewer: ```suggestion
After signing, the returned object contains signed transaction data in a `blob` field, which we can use to submit to the network.
```

To send the above txn we need to first convert it to base64:

```js
const base64Txn = Buffer.from(signedTxnResponse.blob).toString("base64");
const sendTxnResponse = await webmode.sendAndWait(base64Txn);
```

> The sent txn response includes the confirmed transaction response along with txID.

Reviewer: ```suggestion
The `sendTxnResponse` includes the confirmed transaction response along with txID.
```

> 2. MyAlgo Wallet <br/>
>    We use `connectMyAlgoWallet()` method to connect to My Algo wallet.

Reviewer: ```suggestion
#### MyAlgo Wallet

Let's create `connectMyAlgoWallet()` function to connect to My Algo wallet:
```

```js
   async connectMyAlgoWallet() {
      try {
        let walletURL = {
          token: "",
          server: "https://node.algoexplorerapi.io",
          port: "",
        }
        let myAlgoMode = new MyAlgoWalletSession(walletURL);
        await myAlgoMode.connectToMyAlgo();
        if (myAlgoMode.accounts.length) {
          this.walletAddress = myAlgoMode.accounts[0].address;
        }
      } catch (e) {
        console.error(e);
      }

>    }

Reviewer: why we don't have this function in the `WebMode`?

Author: Do you mean catching the error and throwing it?

Reviewer: nevermind, probably it's no worth

```

MyAlgo Wallet method `signLogicSignatureTxn` also works similar to Algosigner webmode.
To send the signed txn returned from MyAlgo Wallet:

```js
const sendTxnResponse = await algoWallet.sendAndWait(signedTxnResponse.blob);
```

> Here also the sent txn response includes the confirmed transaction response along with txID.

Reviewer: ```suggestion
Here also the `sendTxnResponse` includes the confirmed transaction response along with txID.
```

> 3. Wallet Connect<br/>
>    We use `connectWalletConnect()` method is to connect to Wallet Connect wallet.

Reviewer: ```suggestion
#### Wallet Connect

Let's create  a `connectWalletConnect()` method is to connect to a Wallet Connect session.
```

```js
   async connectWalletConnect() {
      try {
        let walletURL = {
          token: "",
          server: "https://node.algoexplorerapi.io",
          port: "",
        }
        let walletConnector = new WallectConnectSession(walletURL);
        await walletConnector.create(true);
        walletConnector.onConnect((error, response) => {
          if (response.accounts.length) {
            this.walletAddress = response.accounts[0];
          }
        });
      } catch (e) {
        console.error(e);
      }

>    }

Reviewer: again, why we don't have it in the `WebMode`?

> # Algo Builder Web Support for Logic Signature Transaction
> 
> The purpose of this tutorial is to demonstrate the usage of Algo Builder Web Package to sign and send a Logic Signature Transaction using major Algorand wallets (Algo Signer, Wallet Connect, MyAlgoWallet).

Reviewer: ```suggestion
# Algo Builder Web Support for Smart Signature Transaction

The purpose of this tutorial is to demonstrate the usage of Algo Builder Web Package to sign and send a Smart Signature Transaction (lsig) using major Algorand wallets (Algo Signer, Wallet Connect, MyAlgoWallet).
```

### Requirements

- [Algo Builder](https://algobuilder.dev/) and it's [dependencies](https://algobuilder.dev/guide/README#requirements)
- Setup the accounts on [`Algosigner`](https://github.com/PureStake/algosigner), [`Wallet Connect`](https://walletconnect.com/) wallet and [`MyAlgo Wallet`](https://wallet.myalgo.com/). It could be MainNet, TestNet or private net.

## Steps

In this tutorial we will use Algo Builder Wallet Widget to connect to various web wallets. Then we will show how to instrument the wallet to sign a Logic Signature transaction. See [Wallet Widget tutorial](https://developer.algorand.org/solutions/vuejs-multi-wallet-widget/) for more details about the widget.

> ### LogicSignature Transaction

Reviewer: ```suggestion
### Smart Signature Transaction
```
